README.md
# Write-ProgressEx &ndash; extended Write-Progress

[project]:https://github.com/mazzy-ax/Write-ProgressEx
[license]:https://github.com/mazzy-ax/Write-ProgressEx/blob/master/LICENSE
[ps]:https://www.powershellgallery.com/packages/Write-ProgressEx
[nuget]:https://www.nuget.org/packages/Write-ProgressEx
[appveyor]:https://ci.appveyor.com/project/mazzy-ax/write-progressex

[![Build status](https://ci.appveyor.com/api/projects/status/5r91g2yxk74a46mi?svg=true)][appveyor]
[![PowerShell Gallery](https://img.shields.io/powershellgallery/dt/Write-ProgressEx.svg)][ps]
[![NuGet](https://buildstats.info/nuget/Write-ProgressEx)][nuget]
<img src="https://raw.githubusercontent.com/mazzy-ax/Write-ProgressEx/master/Media/Write-ProgressEx-icon.png" align="right" alt="Write-ProgressEx icon">

[Write-ProgressEx][project] extends the functionality of the standard powershell cmdlet. Write-ProgressEx is a powershell advanced function that provides a simple way to show ProgressBars with PercentComplete and SecondsRemaining.

The cmdlet:

* works with pipe;
* works with empty activity string;
* completes all inner progresses if no parameters;
* automatically completes with pipe;
* automatically calculates percents, remaining seconds and elapsed time;
* automatically displays current iteration and totals on progress bar;
* automatically displays info with a progress bar and at the console title;
* automatically set parent id for a inner loop;
* stores totals, current values and actual parameters into the module hashtable;
* provides get/set cmdlets to access actual parameters;
* uses script blocks to show messages with date, time, iterations and elapsed time on events:
  * first iteration;
  * activity changed;
  * status changed;
  * completed.
* provides a counter functional. See [Write-ProgressEx as a counter](Examples/Write-ProgressEx.counter.ps1);
* uses the caller function name or the caller script file name as the Activity;
* accepts `-ShowProgressBar Auto` parameter to reduce the overhead for redrawing a screen. It recognizes `None` and `Force` values also.

Note: the function is not safe with multi-thread.

## Examples

A pipe and a simple loop:

```powershell
$range1 = 1..20
$range1 | Write-ProgressEx "loop 1" -Total $range1.Count -ShowMessages | ForEach-Object {
    # ....
}

$range2 = 1..15
$range2 | ForEach-Object {
    # ....
    Write-ProgressEx "loop 2" -Total $range2.Count -Increment
}

Write-ProgressEx #close all progress bars
```

Pipes in nested loops:

```powershell
$outer = 1..20
$inner = 1..50

$outer | Write-ProgressEx "pipe nodes" -Status "outer" -Total $outer.Count -ShowMessages | ForEach-Object {
    $inner | Write-ProgressEx "pipe names" -id 1 -Status "inner" -Total $inner.Count | ForEach-Object {
        # ....
    }
}
```

![screenshot: Write-ProgressEx](Media/examples.pipe.png)

![screenshot: Result messages](Media/examples.messages.png)

A long time command:

```powershell
$path = $env:homepath
$files = Get-ChildItem $path -Recurse | Write-ProgressEx $path
```

More samples are in the folder [Examples](Examples).

## Installation

Automatic install Write-ProgressEx module from the [PowerShell Gallery][ps]:

```powershell
Install-Module -Name Write-ProgressEx
Import-Module Write-ProgressEx
```

Automatic install Write-ProgressEx module from the [NuGet.org][nuget]:

```powershell
Install-Package -Name Write-ProgressEx
Import-Module Write-ProgressEx
```

or manual:

* Download and unblock the latest .zip file.
* Extract the .zip into your `$PSModulePath`, e.g. `~\Documents\WindowsPowerShell\Modules`.
* Ensure the extracted folder is named `Write-ProgressEx`.
* Set an execution policy to `RemoteSigned` or `Unrestricted` to execute not signed modules `Set-ExecutionPolicy RemoteSigned`.
* Run `Import-Module Write-ProgressEx`.

## Known issues

* Unable to import module on case-sensitive file systems. Thanks @jasonchester for the [workaround](https://github.com/mazzy-ax/Write-ProgressEx/issues/5).

## Changelog

* [CHANGELOG.md](CHANGELOG.md)
* <https://github.com/mazzy-ax/Write-ProgressEx/releases>.

## License

This project is released under the [licensed under the MIT License][license].

mazzy@mazzy.ru

CHANGELOG.md
# Changelog

All notable changes to the [project](https://github.com/mazzy-ax/Write-ProgressEx) will be documented in this file.
See also <https://github.com/mazzy-ax/Write-ProgressEx/releases>.

## [v0.22](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.21...v0.22) - 2020-04-26

* Type of parameters `Current` and `Total` changed to `[long]` from `[int]`

## [v0.21](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.20...v0.21) - 2019-03-28

* parameters added:
  * `Remaining`
  * `ShowConsoleTitle`
  * `ShowElapsed`
* features improved:
  * UpdateInterval, Reset, Elapsed, Remaining, PercentComplete, SecondsRemaining
  * `auto name` feature works with functions and script names
* performance improved
* examples improved
  * param `delayMS` added to examples. The default value (30 ms) is used when the user runs this script, 0 is used in unit tests
  * `Write-ProgressEx` with empty arguments added to the begin of each example. It fixes problems when the user interrupts the example and runs another example
  * `-ShowConsoleTitle` parameter added to the pipe example
* style fixed:
  * trail spaces removed due style check
  * for case-sensitive systems, the name of methods and properties uses same case

## [v0.20](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.19...v0.20) - 2018-05-13

* [System.Diagnostics.Stopwatch]$Stopwatch replaced on [DateTime]$StartDateTime and [TimeSpan]$Elapsed
* parameters removed:
  * `NoProgressBar`
* parameters added:
  * `ShowProgressBar` with `Auto`, `Force` and `None` values
  * `StartDateTime`
  * `ProgressDateTime`
  * `Elapsed`
  * `UpdateDateTime`
  * `UpdateInterval`
* fixed due changed parameters:
  * Get-ProgressEx, Set-ProgressEx, Write-ProgressEx cmdlets
  * examples
  * messages
* reamde and comments updated

## [v0.19](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.18...v0.19) - 2018-05-06

* The switch `Reset` added to `Write-ProgressEx` cmdlet.
* Deep refactoring in the cmdlet `Set-ProgressEx`.
* `On completed message` clarified.
* The directive `#require -module Write-ProgressEx` added to example scripts.
* Splatting example added.
* Examples and tests cleanup.
* Readme changed.

## [v0.18](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.17...v0.18) - 2018-05-06

* The function `Set-ProgressEx` fixed.
* The example `Write-ProgressEx.autoName.ps1` fixed.
* The cmdlet `Set-StrictMode` removed from project scripts.
* The directive `#require -version 3.0` removed from project scripts.
* The function `nz` extracted from `Write-ProgressExMessage`.
* The `chocolateyInstall.ps1` removed.
* The directory structure reorganized to remove media, examples and tests from nuget downloads and powershell gallery.
* The project meta info tests added.
* Tests fixed.
* Readme changed.
* Typo corrected.

## [v0.17](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.16...v0.17) - 2017-11-06

* The main goal is a code housekeeping and refactoring
* The internal function Write-ProgressExStd removed. It was not exported function.
* It has refactoring of Cmdlets
  * Set-ProgressEx show standard progress bar as is. This cmdlet does not make a changing of parameters.
  * Write-Progress maintains parameters and switches. It calculates RemainSeconds, PercentComplete, increment Current value, passthru InputValue. It call Set-ProgressEx to show standard progress bar and messages.

## [v0.16](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.15...v0.16) - 2017-10-27

* The cmdlet uses an array of scriptblocks for each event instead a scriptblock
* NoProgressBar error fixed

## [v0.15](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.14...v0.15) - 2017-10-22

* Message scriptblocks should display message instead return message string
* The region block for module variables added

## [v0.14](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.13...v0.14) - 2017-10-21

* The autoname functionality added: If user does not specify an Activity parameter, the Write-ProgressEx display the caller function name or the caller script file name as the Activity.
* A pipe with empty arg list fixed
* Events renamed

## [v0.13](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.12...v0.13) - 2017-10-14

### Added

* Example 'counter' demonstrate how to use [Write-ProgressEx cmdlet as counter](Examples/Write-ProgressEx.counter.ps1).
* Tests with tag build added. It need to run and test examples.

### Changed

* Message templates are scriptblock now
* Cmdlet show warning messages instead write to a host
* The total grow if the current pass over the total
* The total parameter is integer only. Yes, again. Sorry, a runtime checking is not comfortable for XML, files and other iterable ranges
* chocolateyInstall.ps1 moved to the [Tools](Tools/) directory

## [v0.12.1](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.12...v0.12.1) - 2017-10-10

* SecondsRemaining calculation fixed
* Screenshot with Result text fixed

## [v0.12](https://github.com/mazzy-ax/Write-ProgressEx/compare/v0.11...v0.12) - 2017-10-10

### Main goal

* Provide information about total iterations, elapsed time, start time, end time
* Provide template functionality to change information messages

### Added

* The ShowMessages switch
* The in-module cmdlet Write-ProgressExMessage to show event messages
* Event Messages: onBegin, onNewActivity, onNewStatus, onEnd.
* The NoProgressBar switch
* The in-module cmdlet Write-ProgressExStd to call standard cmdlet Write-Progress
* The Total parameter accept a integer and an array

### Changed

* The Current can be more then Total. It's used to make iterations when Total is not known.
* The Stopwatch redesigned: stopwatch start with first iterations. The messages uses the stopwatch to display a total elapsed time
* The module icon
* Examples and Readme

### Removed

* Iterations parameter
* Export module members in Write-ProgressEx.psm1. Module members exported by psd1-file only


Write-ProgressEx.Tests.ps1
# mazzy@mazzy.ru, 2017-11-06
# https://github.com/mazzy-ax/Write-progressEx

#requires -version 3.0
Set-StrictMode -Version 1.0

$me = Split-Path -Leaf $PSCommandPath
$path = Split-Path -Parent $PSCommandPath
$module = Join-Path $path "Write-ProgressEx.psd1"
Import-Module -Force $module

Describe "Write-ProgressEx" -Tag "Run", "UnitTest", "UT" {
    $outerLevel = 1..3
    $innerLevel = 1..2

    Context "Work" {
        It "work" {
            { Write-ProgressEx } | Should not throw
        }

        It "double trouble work" {
            { Write-ProgressEx; Write-ProgressEx } | Should not throw
        }

        It "parameters: outer and inner loops" {
            {
                Write-ProgressEx "outer" -Total $outerLevel.Count
                $outerLevel | ForEach-Object {
                    Write-ProgressEx "inner" -Total $innerLevel.Count -id 1
                    $innerLevel | ForEach-Object {
                        #start-sleep 1
                        Write-ProgressEx -id 1 -increment
                    }
                    Write-ProgressEx -increment
                }
                Write-ProgressEx -Completed
            } | Should not throw
        }

        It "pipes: outer and inner loops" {
            {
                $outerLevel | Write-ProgressEx "outer" -Total $outerLevel.Count | ForEach-Object {
                    $innerLevel | Write-ProgressEx "inner" -Total $innerLevel.Count -id 1 | ForEach-Object {
                        # ...
                    }
                }
            } | Should not throw
        }

        It "progress info: splatting" {
            $pInfo = Get-ProgressEx -Force
            $pInfo.Activity = 'splating'
            $pInfo.Status = 'another status'
            $pInfo.Total = $outerLevel.Count

            { Write-ProgressEx @pInfo } | Should not throw

            Write-ProgressEx
        }
    }

    Context "Functional" {
        AfterEach {
            Write-ProgressEx # complete all
        }

        It "store paramters in module variable" {
            Write-ProgressEx -Total $outerLevel.Count
            (Get-ProgressEx).Total | Should be $outerLevel.Count
        }

        It "increment counter after the increment" {
            Write-ProgressEx -Total $outerLevel.Count
            (Get-ProgressEx).current | Should be 0

            Write-ProgressEx -increment
            (Get-ProgressEx).current | Should be 1
        }

        It "autocomplete with pipe" {
            $outerLevel | Write-ProgressEx "outer" -Total $outerLevel.Count | ForEach-Object {
                $innerLevel | Write-ProgressEx "inner" -Total $innerLevel.Count -id 1 | ForEach-Object {
                    # ...
                    (Get-ProgressEx -id 0).Total | Should be $outerLevel.Count
                    (Get-ProgressEx -id 1).Total | Should be $innerLevel.Count
                    break
                }
                (Get-ProgressEx -id 0).Total | Should be $outerLevel.Count
                (Get-ProgressEx -id 1).Total | Should be $null
                break
            }
            (Get-ProgressEx -id 0) | Should be $null
        }

        It "complete children progress when 'previous' id used" {
            Write-ProgressEx -Total $outerLevel.Count
            Write-ProgressEx -Total $innerLevel.Count -id 1
            (Get-ProgressEx -id 0).current | Should be 0
            (Get-ProgressEx -id 1).current | Should be 0

            Write-ProgressEx -id 1 -increment
            (Get-ProgressEx -id 0).current | Should be 0
            (Get-ProgressEx -id 1).current | Should be 1

            Write-ProgressEx -id 0 -increment
            (Get-ProgressEx -id 0).current | Should be 1
            (Get-ProgressEx -id 1) | Should be $null

            Write-ProgressEx
            (Get-ProgressEx -id 0) | Should be $null
        }

        It "without parameters complete all children progresses" {
            Write-ProgressEx "outerLevel" -Total $outerLevel.Count
            Write-ProgressEx "innerLevel" -Total $innerLevel.Count -id 1

            (Get-ProgressEx -id 0).Total | Should be $outerLevel.Count
            (Get-ProgressEx -id 1).Total | Should be $innerLevel.Count

            Write-ProgressEx
            (Get-ProgressEx) | Should be $null
            (Get-ProgressEx -id 1) | Should be $null
        }
    }
}

Describe "Get-ProgressEx" {

    Context "work" {
        It "get" {
            { Get-ProgressEx } | Should not throw
        }

        It "get -force" {
            { Get-ProgressEx -force } | Should not throw
        }
    }

    Context "functional" {
        It "name parameter" {
            $pInfo = Get-ProgressEx -id 1
            $pInfo | Should be $null
        }

        It "name parameter -force" {
            $pInfo = Get-ProgressEx -id 1 -force
            $pInfo.Id | Should be 1
        }

        It "position parameter" {
            $pInfo = Get-ProgressEx 1
            $pInfo | Should be $null
        }

        It "position parameter -force" {
            $pInfo = Get-ProgressEx 1 -force
            $pInfo.Id | Should be 1
        }

        It "pipe parameter" {
            $pInfo = 1 | Get-ProgressEx
            $pInfo | Should be $null
        }

        It "pipe parameter -force" {
            $pInfo = 1 | Get-ProgressEx -force
            $pInfo.Id | Should be 1
        }

        It "pipe array parameter" {
            $pInfo = 1, 30, 40 | Get-ProgressEx
            $pInfo.Count | Should be 3
            $pInfo[0] | Should be $null
            $pInfo[1] | Should be $null
            $pInfo[2] | Should be $null
        }

        It "pipe array parameter -force" {
            $pInfo = 1, 30, 40 | Get-ProgressEx -force
            $pInfo.Count | Should be 3
            $pInfo[0].Id | Should be 1
            $pInfo[1].Id | Should be 30
            $pInfo[2].Id | Should be 40
        }
    }
}

Describe "Set-ProgressEx" {

    Context "work" {
        It "work" {
            { Set-ProgressEx } | Should not throw
        }

        It "double trouble work" {
            { Set-ProgressEx; Set-ProgressEx } | Should not throw
        }
    }
}

<#
#remove-module 'Write-ProgressEx'

Describe "Write-ProgressExMesage" {

    InModuleScope {
        Context "work" {
            $Message = {"Test"}
            $pInfo = @{
                id                       = 1
                ShowMessageOnBegin       = $true
                ShowMessageOnNewActivity = $true
                ShowMessageOnNewStatus   = $true
                ShowMessageOnEnd         = $true
                MessageOnBegin           = {"$($pInfo.id):onBegin"}
                MessageOnNewActivity     = {"$($pInfo.id):onNewActivity"}
                MessageOnNewStatus       = {"$($pInfo.id):onNewStatus"}
                MessageOnEnd             = {"$($pInfo.id):onEnd"}
            }
            $Expected = @{
                Message              = Invoke-Command $Message
                MessageOnBegin       = Invoke-Command $pInfo.MessageOnBegin
                MessageOnNewActivity = Invoke-Command $pInfo.MessageOnNewActivity
                MessageOnNewStatus   = Invoke-Command $pInfo.MessageOnNewStatus
                MessageOnEnd         = Invoke-Command $pInfo.MessageOnEnd
            }

            It "message" {
                Write-ProgressExMessage $pInfo -Message $Message | Should be $Expected.Message
            }

            It "onBegin" {
                Write-ProgressExMessage $pInfo -ShowMessagesOnBegin | Should be $Expected.MessageOnBegin
            }

            It "onNewActivity" {
                Write-ProgressExMessage $pInfo -ShowMessagesOnNewActivity | Should be $Expected.MessageOnNewActivity
            }

            It "onNewStatus" {
                Write-ProgressExMessage $pInfo -ShowMessagesOnNewStatus | Should be $Expected.MessageOnNewStatus
            }

            It "onEnd" {
                Write-ProgressExMessage $pInfo -ShowMessagesOnEnd | Should be $Expected.MessageOnEnd
            }
        }
    }
}
#>

<#
Describe "Write-ProgressExStd" {

    Context "work" {
        It "not exported" {
            { get-command Write-ProgressExStd -Module Write-ProgressEx } | Should throw
        }

        InModuleScope Write-ProgressEx {
            It "by pInfo" {
                $pInfo = Get-ProgressEx -id 0
                { Write-ProgressExStd -pInfo $pInfo } | Should not throw
            }
        }
    }
}
#>




